Payment Intent and Checkout Session (SCA, EU Regulations)

tiwariayush opened the pull request:
Add models and functionalities to support SCA regulations (https://stripe.com/docs/strong-customer-authentication/migration)

New Models:
- Payment Method
- Setup Intent
- Payment Intent
- Checkout

Also, extra changes to support model additions such as:
- Update Subscriptions with new `pending_setup_intent` field 
- Rename property payment_methods for Customer object to solve conflict: Due to addition of new Payment Method model, there has been a conflict in the name of property `payment_methods` on Customer object, hence the required name change.
- Update migrations and test fixtures.
-  Add `Charge.payment_intent` + test fixture

This resolves #803, #897

aymericderbois committed:
Add payment method and payment intent models

Also,
- Add new signals for payment intent
- Update event handlers to add payment method and intent
- Create admins for the aforementioned models

Co-authored-by: Ayush Tiwari <ayush.tiwari@bureauxlocaux.com>

therefromhere commented on djstripe/models/payment_methods.py:
```diff
@@ -455,3 +455,41 @@ def detach(self):
+	stripe_dashboard_item_name = "payment methods"
+
+	@classmethod
+	def attach(cls, payment_method_id, stripe_customer, api_key=djstripe_settings.STRIPE_SECRET_KEY):
```
I don't think api_key parameter is needed here?

tiwariayush committed:
Add migrations for new models and rename old indexes

tiwariayush committed:
Rename property payment_methods for Customer object to solve conflict

Due to addition of new Pyament Method model, there has been a conflict
in the name of property `payment_methods` on Custome object, hence the
name change.

therefromhere commented on djstripe/migrations/0005_auto_20190703_0958.py:
```diff
@@ -0,0 +1,88 @@
+		migrations.AlterField(
+            model_name='payout',
+            name='status',
+            field=djstripe.fields.StripeEnumField(enum=djstripe.enums.PayoutStatus, help_text='Current status of the payout. A payout will be `pending` until it is submitted to the bank, at which point it becomes `in_transit`. It will then change to paid if the transaction goes through. If it does not go through successfully, its status will change to `failed` or `canceled`.', max_length=10),
```
If this alterfield is just changing help_text it can be done by editting the original migration 

(Not a huge problem here since we're adding a migration anyway, but it means typo fixes can potentially be backported to stable branches)

therefromhere commented on djstripe/models/core.py:
```diff
@@ -466,7 +466,7 @@ def credits(self):
 
 	@property
-	def payment_methods(self):
+	def customer_payment_methods(self):
```
Mention this change in HISTORY.rst as a breaking change

kavdev replied:
(this will require a minor version change)

therefromhere reviewed (commented):
Looks like you're making good progress, good catch on the index rename.

Can you also add update HISTORY.rst to your to-do (at least a first pass, I can tidy up if needed).

tiwariayush committed:
Update HISTORY.rst

tiwariayush reviewed (commented):
An old commit which can be mentioned in commit message here: 34a581fde

tiwariayush committed:
Add Checkout model and signals

tiwariayush reviewed (commented):
@therefromhere Hi, I think this commit (6d7b89dbf16) can go in master outside of this PR. Should I create separate PR for this ?

tiwariayush committed:
Merge migrations

tiwariayush committed:
Udpate Subscriptions with new `pending_setup_intent` field

Also, update test fixtures and tests.

tiwariayush committed:
Example for payment intent

kavdev commented on djstripe/models/checkout.py:
```diff
@@ -0,0 +1,117 @@
+			"The ID of the PaymentIntent created if SKUs or line items were provided."
+		)
+	)
+	payment_method_types = JSONField(
```
Could this be an enum field?

tiwariayush replied:
I'm not sure if this should be an Enum field as it is supposed to be an array of payment method types.

kavdev commented on djstripe/models/checkout.py:
```diff
@@ -0,0 +1,117 @@
+		)
+	)
+	# Should be Enum field ??
+	submit_type = models.CharField(
```
Could this be an enum field?

tiwariayush replied:
Fixed

kavdev commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1224,331 @@ def is_valid_object(cls, data):
+			"fraudulent, requested_by_customer, or failed_invoice."
+		)
+	)
+	capture_method = models.CharField(
```
Could this be an enum field?

tiwariayush replied:
Done

kavdev commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1224,331 @@ def is_valid_object(cls, data):
+			"publishable key."
+		)
+	)
+	confirmation_method = models.CharField(
```
Could this be an enum field?

tiwariayush replied:
Done

kavdev commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1224,331 @@ def is_valid_object(cls, data):
+			"ID of the payment method used in this PaymentIntent."
+		)
+	)
+	payment_method_types = JSONField(
```
Could this be an enum field?

tiwariayush replied:
Same as one for `Session` class, this is supposed to be array of payment types, hence I left it as it is.

kavdev commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1224,331 @@ def is_valid_object(cls, data):
+	)
+	# XXX: Should we use EnumField for this one as it is similar to SetupIntent.usage where we have same
+	# options `on_session` and `off_session` ?
+	setup_future_usage = models.CharField(
```
Could this be an enum field?

tiwariayush replied:
Done

kavdev commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1224,331 @@ def is_valid_object(cls, data):
+	stripe_class = stripe.SetupIntent
+	stripe_dashboard_item_name = "setup intents"
+
+	application = models.CharField(
```
Could this be a ForeignKey + new model?

tiwariayush replied:
I am confused on this one about the mention of `Connect Application` [ in the API here ](https://stripe.com/docs/api/setup_intents/object#setup_intent_object-application) as I couldn't find mention of `Application` in the API elsewhere. Help on this appreciated :)

therefromhere replied:
Maybe leave it un-implemented for now with a TODO, if we're not sure?

tiwariayush replied:
Fixed in 2965902713

lx-stripe replied:
Impressive work!! Congrats!! re. this setup intent field, it has the same meaning as `payment_intent.application` [[api-ref]](https://stripe.com/docs/api/payment_intents/object#payment_intent_object-application): the value is the unique identifier assigned to the _Connect_ application. It's, for instance, the `client_id` informed in `/oauth/authorize` [docs](https://stripe.com/docs/connect/oauth-reference#get-authorize)

therefromhere replied:
@lx-stripe ah, so we can treat it as an opaque string that doesn't correspond to an API object?

lx-stripe replied:
yup... technically it belongs to the account but, since there's no way to retrieve the application client id over the API, an opaque string would be the right choice imho

therefromhere replied:
OK @tiwariayush we don't need to TODO for this one then

kavdev commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1224,331 @@ def is_valid_object(cls, data):
+			"ID of the payment method used in this PaymentIntent."
+		)
+	)
+	payment_method_types = JSONField(
```
Could this be an enum field?

tiwariayush replied:
^^ Array type

kavdev reviewed (changes requested):
Solid work @askpatrickw and @tiwariayush! A few comments on some field choices, but other than that, lgtm. I'm assuming the example app works as expected.

tiwariayush committed:
Add migrations

tiwariayush committed:
Add SetupIntent model

- Also, add migrations.
- Update Setup Intent webhook signals
- Add admin for setup intent

tiwariayush committed:
Add migrations

tiwariayush committed:
fixup! Example for payment intent

tiwariayush committed:
fixup! Update HISTORY.rst

therefromhere commented on tests/apps/example/templates/payment_intent.html:
```diff
@@ -0,0 +1,179 @@
+			errorElement.textContent = result.error.message;
+		} else {
+		  // Otherwise send paymentMethod.id to your server.
+		  fetch('/example/payment-intent', {
```
Can you use {% url %} here

tiwariayush replied:
Fixed in 1f544bc62

therefromhere commented on tests/apps/example/templates/payment_intent.html:
```diff
@@ -0,0 +1,179 @@
+		  } else {
+			// The card action has been handled
+			// The PaymentIntent can be confirmed again on the server
+			fetch('/ajax/confirm_payment', {
```
Shouldn't this  point to a view?

tiwariayush replied:
Fixed in 1f544bc62

therefromhere reviewed (changes requested):
Cool, just had a quick look on the phone, I'll try and give it a proper run-through this weekend

kavdev commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1224,330 @@ def is_valid_object(cls, data):
+			"ID of the review associated with this PaymentIntent, if any."
+		)
+	)
+	# XXX: Should we use EnumField for this one as it is similar to SetupIntent.usage where we have same
```
you can remove this comment. also, idk if setting a default here is a good idea

tiwariayush replied:
Fixed in 7704be8

therefromhere commented on djstripe/migrations/0005_auto_20190703_0958.py:
```diff
@@ -0,0 +1,83 @@
+		# better to rename the old index.
+		migrations.RunSQL(
+			'ALTER INDEX djstripe_paymentmethod_id_0b9251df_like rename TO djstripe_paymentmethod_legacy_id_0b9251df_like',
+			'ALTER INDEX djstripe_paymentmethod_legacy_id_0b9251df_like rename TO djstripe_paymentmethod_id_0b9251df_like',
```
Ah, I think this command is postgres specific, I'll see if I can change it to work on mysql (and hopefully sqlite too).

therefromhere replied:
done in https://github.com/dj-stripe/dj-stripe/pull/914/commits/d56dc5c05a84c25bc9f8f51c2baf7595a3fdce0e

therefromhere commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1224,330 @@ def is_valid_object(cls, data):
+			"An arbitrary string attached to the object. Often useful for displaying to users."
+		)
+	)
+	invoice = models.ForeignKey(
```
I think this maybe should be a OneToOneField?

And maybe it belong on Invoice rather than here, since the PI is created from the invoice?

> The PaymentIntent associated with this invoice. The PaymentIntent is generated when the invoice is finalized, and can then be used to pay the invoice. Note that voiding an invoice will cancel the PaymentIntent.

See https://stripe.com/docs/api/invoices/object#invoice_object-payment_intent / 
https://stripe.com/docs/api/payment_intents/object#payment_intent_object-invoice

tiwariayush replied:
Fixed in 7704be8c6

Will add migrations after cleaning up the commits(Added in TODO for PR)

therefromhere commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1224,330 @@ def is_valid_object(cls, data):
+			"Email address that the receipt for the resulting payment will be sent to."
+		)
+	)
+	review = models.CharField(
```
I'm pretty sure this should be a ForeignKey to a new Review model - remove this for now and put a TODO?

(otherwise migration might be a pain when we do add the field)

therefromhere replied:
#670 refers to that, I think

tiwariayush replied:
Fixed in 7704be8c6

therefromhere commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1224,330 @@ def is_valid_object(cls, data):
+		on_delete=models.SET_NULL,
+		null=True,
+		help_text=(
+			"ID of the payment method used in this PaymentIntent."
```
Since this is a ForeignKey, maybe the help_text shouldn't mention ID?

eg

"Payment method used in this PaymentIntent."

instead?

therefromhere replied:
(same for all of the new fks)

tiwariayush replied:
Fixed in 7704be8c6, 313d8b64 and 8821bd7d5

therefromhere committed:
Add Charge.payment_intent + test fixture

therefromhere committed:
Make migration compatible with other dbs (mysql/sqlite)

therefromhere commented on djstripe/models/core.py:
```diff
@@ -759,6 +759,12 @@ def add_card(self, source, set_default=True):
+		from .payment_methods import PaymentMethod
+
+		stripe_customer = self.api_retrieve()
+		PaymentMethod.attach(payment_method, stripe_customer)
```
Could do with test coverage of this

tiwariayush replied:
I have added fixtures for payment method. Not sure if its cent percent correct way 171a500f5. I am still bit confused how to use it in tests though. That's why I haven't been able to add coverage for this.

lx-stripe replied:
@tiwariayush I'd suggest using a test token instead of a test card (something like [this](https://runkit.com/lxmedina/5d40210dc16a86001358f2d6)); otherwise the tests would fail if the account has not activated [Process payments unsafely](https://dashboard.stripe.com/account/integration/settings). Other option would be to use test payment methods directly (3rd column [here](https://stripe.com/docs/testing#cards))

therefromhere replied:
@tiwariayush cool it looks like you're on the right track, I think it just needs a tweak for the magic id faking.   I'll try taking a look at that today.

therefromhere replied:
Fixture update done in ac4d439da0d43c07990b6808e2f36a465bbf38a6 , and I've added a test in 3c8bb780c99d009e5d9dad51e3af8f88ef7da6df

therefromhere replied:
also another test in 48993f2cb06e4a6d9906ce4fa65e1b695f16afe3

tiwariayush committed:
fixup! Add SetupIntent model

therefromhere committed:
Fix typo

Found with card 4000 0025 0000 3155 as per https://stripe.com/docs/payments/payment-intents/migration#testing

therefromhere commented on tests/apps/example/templates/payment_intent.html:
```diff
@@ -152,7 +152,7 @@ <h1 class="is-size-3">Example Payment Intent Manual Configuration</h1>
 	  } else if (response.requires_action) {
 		// Use Stripe.js to handle required card action
 		stripe.handleCardAction(
-		  response.payment_intent_client_secretemail
```
FYI @tiwariayush

tiwariayush replied:
Ah thanks for noticing. I didn't check the implementation of example with other cards than the success one (4242).

therefromhere replied:
Note that it still fails at the next step since it's trying to call `ajax/confirm_payment`

tiwariayush replied:
Ah yes, I'm fixing the template currently. `ajax/confirm_payment` should point to the same view `example/payment-intent`

tiwariayush replied:
Fixed in cc9bfa228

tiwariayush committed:
fixup! Add payment method and payment intent models

tiwariayush committed:
fixup! Add SetupIntent model

tiwariayush committed:
fixup! Add Checkout model and signals

tiwariayush committed:
fixup! Add payment method and payment intent models

tiwariayush committed:
fixup! Example for payment intent

tiwariayush committed:
fixup! Example for payment intent

tiwariayush committed:
Add migrations

tiwariayush committed:
Update migrations

tiwariayush committed:
Update tests to include empty foreign key

tiwariayush committed:
fixup! Update migrations

tiwariayush committed:
fixup! Add payment method and payment intent models

tiwariayush committed:
Add fixtures for Payment Method

tiwariayush committed:
fixup! Add fixtures for Payment Method

tiwariayush committed:
fixup! Add fixtures for Payment Method

tiwariayush committed:
Code reformating

therefromhere committed:
Fix payment method fixture creation

- attach to a customer, and save so metadata is saved in stripe

therefromhere committed:
remove unnecessary code

therefromhere committed:
fix bad commit

therefromhere committed:
regenerated (nulled fields we're not setting, newline at end)

therefromhere commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1237,285 @@ def is_valid_object(cls, data):
+	stripe_class = stripe.SetupIntent
+	stripe_dashboard_item_name = "setup intents"
+
+	# TODO: `application` field would go better as Foreign Key. Should be changed after adding new model.
```
remove TODO as per discussion below

therefromhere committed:
Added test of PaymentMethod.attach

changed implementation to use stripe_customer["id"] instead of .id so it works with our test dict (real object allows attr or dict access)
also added api_key

therefromhere committed:
Reformat

therefromhere committed:
Added simple test of Customer.add_payment_method

therefromhere committed:
Removed uneeded TODO

see https://github.com/dj-stripe/dj-stripe/pull/914#discussion_r309142401

therefromhere commented on djstripe/models/payment_methods.py:
```diff
@@ -455,3 +455,60 @@ def detach(self):
+		stripe_payment_method = stripe.PaymentMethod.attach(
+			payment_method_id, customer=stripe_customer["id"], api_key=api_key
+		)
+		return cls._create_from_stripe_object(stripe_payment_method)
```
I need to test, but I think this should use sync_from_stripe_data instead of _create_from_stripe_object, since the payment_method may already exist in the DB.

therefromhere replied:
resolved in ba608b4c9cfee098da6e54cd4658022d74f7fdaa

therefromhere committed:
Use sync_from_stripe_data in PaymentMethod.attach

so we support adding PaymentMethod that was synced before it was attached

therefromhere commented on djstripe/models/core.py:
```diff
@@ -1218,6 +1239,284 @@ def is_valid_object(cls, data):
+	stripe_class = stripe.PaymentIntent
+	stripe_dashboard_item_name = "payment intents"
+
+	amount = StripeQuantumCurrencyAmountField(
```
@kavdev is StripeQuantumCurrencyAmountField the right type to use here?  (as opposed to StripeDecimalCurrencyAmountField).

See #247
